BenefitInliner phase 1/3: Basic classes for performing abstract interpretation (#5507)
eclipse/omr: doc/compiler/optimizer/BenefitInliner.md

# BenefitInliner

BenefitInliner introduces a novel way of inlining. Compared to the traditional way of inlining relies on hard-coded heuristics, BenefitInliner provides a systematic way of evaluating inlining candidates that combines the dynamic and static benefits of inlining. 

## Overview of main phases of BenefitInliner
1. Build the search space for candidate methods to be inlined.
    1. **IDTBuilder** constructs an inlining dependency tree (**IDT**).
    2. **AbstractInterpreter** generates an **InliningMethodSummary** stashing potential optimization opportunities and specifying the constraints for parameters to make optimizations happen.

>     3. **AbstractInterpreter** computes abstract state values for each program state. 

Reviewer: I don't understand this sentence. Is "each program state" supposed to be "each program point"?

Author: Yes, it should be ""each program point"

    4. At each call site, we calculate the benefit of inlining this method with its call frequency (direct benefit) and optimizations unlocked (indirect benefit) according to **InliningMethodSummary**.
2. Run the knapsack algorithm for choosing the best inlining plan.  
3. Commit the inlining plan and inline the methods.
    
## Higher-Level Structure of the parts of BenfitInliner

### 1. BenefitInliner.cpp

Class hierarchy of the adapter class:
* **TR::Optimization**
    * **BenefitInlinerWrapper**

Class hierarchy of the real inliner:
* **TR_InlinerBase**
    * **BenefitInlinerBase**
       * **BenefitInliner**

* **BenefitInlinerWrapper** is an adapter class which extends from **TR::Optimization**. `BenefitInlinerWrapper::perform()` initializes a **BenefitInliner** and controls how benefit inlining process is being done.

> * **BenefitInliner** is the real Inliner that contains functionalities for selecting inlining candidates. When a **BenefitInliner** is initialized, it comes with an inlining budget that constrains inlining by avoiding an uncontrollable increase in binary size. The inlining budget will be calculated based on the hotness and size of the method being JIT compiled. `BenefitInliner::buildingInliningDependecyTree()` builds an **IDT** that contains all the candidate methods to be inlined. `BenefitInliner::inlinerPacking()` consumes the **IDT** and selects the optimal set of methods to be inlined given the inlining budget and the cost and benefit of each method in the **IDT**. 

Reviewer: ```suggestion
* **BenefitInliner** is the real Inliner that contains functionalities for selecting inlining candidates. When a **BenefitInliner** is initialized, it comes with an inlining budget that constrains inlining by avoiding an unbounded increase in binary size. The inlining budget will be calculated based on the hotness and size of the method being JIT compiled. `BenefitInliner::buildingInliningDependecyTree()` builds an **IDT** that contains all the candidate methods to be inlined. `BenefitInliner::inlinerPacking()` consumes the **IDT** and selects the optimal set of methods to be inlined given the inlining budget and the cost and benefit of each method in the **IDT**. 
```

* **BenefitInlinerBase** contains the functionalities of doing the actual inlining according to the inlining proposal proposed by `BenefitInliner::inlinerPacking()`. 

### 2. InliningProposal.cpp

* **InliningProposal** proposes an optimal set of methods that will be inlined. **BenefitInlinerBase** takes **InliningProposal** as a dependency. **InliningProposal** will be set by the `BenefitInliner::inlinerPacking()` and will be consulted during the inlining. 

> * **InliningProposalTable** is a 2D table of **InliningProposal**. By the nature of `BenefitInliner::inlinerPacking()` being a nested knapsack algorithm, a 2D table is required. 

Reviewer: I thought `InliningProposal` _was_ the table? What exactly is the distinction between `InliningProposal` and `InliningProposalTable`?

Author: The InliningProposalTable is a 2D table of InliningProposal. InliningProposal encapsulates a bit vector which indicates which method should be inlined and which should not. 

By looking at the traditional knapsack algorithm, which selects the optimal set of goods which has the maximum values but does not overweight the weight limits. The algorithm uses a 2D table for dynamic programming computation, an element in the table is a selection plan, which corresponds to the inliningProposal in this case.

### 3. IDTNode.cpp, IDT.cpp & IDTBuilder.cpp

* **IDTNode** is the building block of **IDT**. It is the abstract representation of a candidate method. 

An **IDTNode** has the following important properties:

> > 1. **call ratio** - It estimates how many times this node is visited per the execution of its parent. 
> > 2. **root call ratio** - It estimates how many times this node is visited per the execution of the root node. 
> > 3. **Inlining method summary** - It captures potential optimization opportunities for inlining this method. (Refer to **InliningMethodSummary.cpp** for details)
> > 4. **static benefit** - The indirect benefit (optimizations unlocked) of inlining this method.
> > 5. **budget** - The remaining budget for the descendants of this node.
> > 6. **Bytecode Index** - The call site index of this method.

Reviewer: Minor formatting suggestion
```suggestion
- **call ratio** - Estimate of how many times this node is visited per the execution of its parent. 
- **root call ratio** - Estimate of how many times this node is visited per the execution of the root node. 
- **Inlining method summary** - Captures potential optimization opportunities for inlining this method. (Refer to **InliningMethodSummary.cpp** for details)
- **static benefit** - The indirect benefit (optimizations unlocked) of inlining this method.
- **budget** - The remaining budget for the descendants of this node.
- **Bytecode Index** - The call site index of this method.
```

The **benefit** of inlining a candidate method is calculated as: `Benefit = root call ratio * static benefit`

> The **cost** of inlining a candidate method is equal to the size of this method.

Reviewer: Is this a strict requirement or can a language implementation choose different equations?

Author: This is proposed by the research paper

The **cost** of inlining a candidate method is equal to the size of this method.

The **cost** and **benefit** will be used for solving the inliner packing problem for getting an optimal set of methods to be inlined.

> * **TR::IDT** is an abstract program representation that allows us to consistently define the search space for inlining candidates. The parent-child relationship in the **TR::IDT** corresponds to the caller-caller relationship.

Author: So I can make the method of calculating the benefit as **virtual**

* **IDTBuilder** builds an **IDT** starting from the method (root method) being JIT compiled. The **IDT** is built in **DFS** order. `IDTBuilder::buildIDT()` returns a complete **IDT** that can be consumed by `BenefitInliner::inlinerPacking()`

**Process of building the IDT:**

```
buildIDT() --> buildIDTHelper() --> generateControlFlowGraph() --> performAbstractInterpretation() --> addChild() 
             |                                                                                              |  
             |______________________________________________________________________________________________| 
``` 

Note: `generateControlFlowGraph()` and `performAbstractInterpretation()` need language-specific implementation.

> The responsibilities of the **abstract interpretation**:
> > 1. Identifying call sites. Call sites will be added to the **IDT** as **IDTNodes**
> 
> > 2. Generating inlining method summaries. (Refer to **InliningMethodSummary.cpp** for details )
> 
> > 3. Simulating the program state. Currently, the useful part of the program state is the arguments ( represented by **AbsArguments**) passed to each call site. Combining those arguments with the inlining method summary, we can get the static benefit of inlining this method.

Reviewer: ```suggestion
The responsibilities of **abstract interpretation**:
1. Identifying call sites. Call sites will be added to the **IDT** as **IDTNodes**
2. Generating inlining method summaries. (Refer to **InliningMethodSummary.cpp** for details )
3. Simulating the program state. Currently, the useful part of the program state is the arguments ( represented by **AbsArguments**) passed to each call site. Combining those arguments with the inlining method summary, we can get the static benefit of inlining this method.
```

### 4. AbsVisitor.hpp & IDTBuilderVisitor

> * **AbsVisitor** is an interface for the derived classes to define customized callback methods during abstract interpretation. Currently, there is only one method defined in **AbsVisitor** called `AbsVistor::visitCallSite`. **IDTBuilderVisitor** inherits from **AbsVisitor** and defines it own implementation of `visitCallSite`. Whenever we encounter a call site during abstract interpretation, `AbsVistor::visitCallSite` will be invoked. If **AbsVisitor** is an **IDTBuilderVisitor**, `IDTBuilder::addChild` will be called in order to add an **IDTNode** to the **IDT**.

Reviewer: I'm not sure I understand this description of `AbsVisitor`. What exactly is it for? Maybe a diagram could help?

Author: You will see it in the second phase of contribution. One use case of AbsVisitor is during abstract interpretation, if we encounter a callsite, this callsite will be made into an IDT Node and adding to the IDT. To achieve this, IDTBuilderVisitor inherits from AbsVisitor and implements its own _`visitCallsite`_ method

### 5. InliningMethodSummary.cpp

**PotentialOptimization** is the base class for all the potential optimizations such as branch folding and null-check folding, etc. It takes **VPConstraint** as a dependency to serve as the constraint (lattice). Those constraints specify the maximal safe values to make the optimizations happen. `PotentialOptimization::predicate` tests against an argument value to tell whether or not this abstract argument is a safe value to unlock the optimization.

**InliningMethodSummary** captures potential optimization opportunities of inlining one particular method. `InliningMethodSummary::predicates` tells the total optimizations that will be unlocked given the abstract argument value.

### 6. AbsValue.cpp, AbsStackMachineState.cpp & AbsState.cpp

> **AbsValue** is the abstract representation of value. It uses **VPConstraint** as the lattice to represent the value. A **top** **AbsValue** is the least accurate abstract value, analogous to the unknown value. A **top** **AbsValue** simply has constraint being NULL. `AbsValue::merge` relies on `VPConstraint::merge` for merging various type of values. **AbsValue** comes with a set of factory methods for creating different kinds of abstract values. 

Reviewer: This needs to be updated.

* **AbsState**
    * **AbsStackMachineState**

> **AbsState** is the abstract representation of the program state. **AbsStackMachineState** inherits from **AbsState** and it is used for stack machine language. It takes **AbsLocalVarArray** and **AbsOpStack** as dependencies. **AbsState** is used during abstract interpretation. `TR::Block` takes **AbsState** as an dependency. 

Reviewer: Why does `TR::Block` need to know about `AbsState`?

Author: So the abstract interpretater walks the blocks of the CFG in reverse-post order, interpreting the blocks one by one and merging the absStates of the block's predecessors if it has multiple predecessors by the control flow.  Let TR::Block has an absState is useful.

Reviewer 2: I definitely don't think `TR::Block` should be modified for this work. The abstract interpreter state is only meaningful while doing abstract interpretation and isn't a general property of a block, so I'd much prefer keeping a separate mapping between blocks and states.

Reviewer 3: I agree with @aviansie-ben and @Leonardo2718 on this one.

Author: `AbsState` removed
